File: examples/simple-examples/src/main/java/ai/koog/agents/example/calculator/java/CalculatorTools.java
package ai.koog.agents.example.calculator.java;

import ai.koog.agents.core.tools.annotations.LLMDescription;
import ai.koog.agents.core.tools.annotations.Tool;
import ai.koog.agents.core.tools.reflect.ToolSet;

/**
 * Java implementation of calculator tools, mirroring {@code CalculatorTools.kt}.
 *
 * <p>Demonstrates how to define agent tools using the {@code @Tool} and {@code @LLMDescription}
 * annotations on a class that implements {@link ToolSet}.
 *
 * <p>Note: {@code @LLMDescription} uses {@code description} as its annotation attribute name,
 * so Java callers must use the named form: {@code @LLMDescription("...")}.
 */
@LLMDescription(description = "Tools for basic calculator operations")
public class CalculatorTools implements ToolSet {

    @Tool
    @LLMDescription(description = "Adds two numbers")
    public String plus(
        @LLMDescription(description = "First number") float a,
        @LLMDescription(description = "Second number") float b
    ) {
        return String.valueOf(a + b);
    }

    @Tool
    @LLMDescription(description = "Subtracts the second number from the first")
    public String minus(
        @LLMDescription(description = "First number") float a,
        @LLMDescription(description = "Second number") float b
    ) {
        return String.valueOf(a - b);
    }

    @Tool
    @LLMDescription(description = "Divides the first number by the second")
    public String divide(
        @LLMDescription(description = "First number") float a,
        @LLMDescription(description = "Second number") float b
    ) {
        return String.valueOf(a / b);
    }

    @Tool
    @LLMDescription(description = "Multiplies two numbers")
    public String multiply(
        @LLMDescription(description = "First number") float a,
        @LLMDescription(description = "Second number") float b
    ) {
        return String.valueOf(a * b);
    }
}

File: examples/src/main/kotlin/ai/koog/agents/example/calculator/CalculatorTools.kt
package ai.koog.agents.example.calculator

import ai.koog.agents.core.dsl.builder.forwardTo
import ai.koog.agents.core.dsl.builder.strategy
import ai.koog.agents.core.dsl.extension.nodeExecuteMultipleTools
import ai.koog.agents.core.dsl.extension.nodeLLMCompressHistory
import ai.koog.agents.core.dsl.extension.nodeLLMRequestMultiple
import ai.koog.agents.core.dsl.extension.nodeLLMSendMultipleToolResults
import ai.koog.agents.core.dsl.extension.onAssistantMessage
import ai.koog.agents.core.dsl.extension.onMultipleToolCalls
import ai.koog.agents.core.environment.ReceivedToolResult
import ai.koog.agents.core.tools.annotations.LLMDescription
import ai.koog.agents.core.tools.annotations.Tool
import ai.koog.agents.core.tools.reflect.ToolSet

@Suppress("unused")
@LLMDescription("Tools for basic calculator operations")
class CalculatorTools : ToolSet {

    @Tool
    @LLMDescription("Adds two numbers")
    fun plus(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a + b).toString()
    }

    @Tool
    @LLMDescription("Subtracts the second number from the first")
    fun minus(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a - b).toString()
    }

    @Tool
    @LLMDescription("Divides the first number by the second")
    fun divide(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a / b).toString()
    }

    @Tool
    @LLMDescription("Multiplies two numbers")
    fun multiply(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a * b).toString()
    }
}

object CalculatorStrategy {
    private const val MAX_TOKENS_THRESHOLD = 1000

    val strategy = strategy("test") {
        val nodeCallLLM by nodeLLMRequestMultiple()
        val nodeExecuteToolMultiple by nodeExecuteMultipleTools(parallelTools = true)
        val nodeSendToolResultMultiple by nodeLLMSendMultipleToolResults()
        val nodeCompressHistory by nodeLLMCompressHistory<List<ReceivedToolResult>>()

        edge(nodeStart forwardTo nodeCallLLM)

        edge(
            (nodeCallLLM forwardTo nodeFinish)
                    transformed { it.first() }
                    onAssistantMessage { true }
        )

        edge(
            (nodeCallLLM forwardTo nodeExecuteToolMultiple)
                    onMultipleToolCalls { true }
        )

        edge(
            (nodeExecuteToolMultiple forwardTo nodeCompressHistory)
                    onCondition { llm.readSession { prompt.latestTokenUsage > MAX_TOKENS_THRESHOLD } }
        )

        edge(nodeCompressHistory forwardTo nodeSendToolResultMultiple)

        edge(
            (nodeExecuteToolMultiple forwardTo nodeSendToolResultMultiple)
                    onCondition { llm.readSession { prompt.latestTokenUsage <= MAX_TOKENS_THRESHOLD } }
        )

        edge(
            (nodeSendToolResultMultiple forwardTo nodeExecuteToolMultiple)
                    onMultipleToolCalls { true }
        )

        edge(
            (nodeSendToolResultMultiple forwardTo nodeFinish)
                    transformed { it.first() }
                    onAssistantMessage { true }
        )
    }
}